src/parser/Parser.ts

```diff
@@ -174,6 +174,17 @@ export class Parser {
             } else {
                 name = consume(`Expected ${functionType.text} name after '${functionType.text}'`, Lexeme.Identifier) as Identifier;
                 leftParen = consume(`Expected '(' after ${functionType.text} name`, Lexeme.LeftParen);
+
+                //prevent functions from ending with type designators
+                let lastChar = name.text[name.text.length - 1];
+                if (["$", "%", "!", "#"].indexOf(lastChar) > -1) {
+                    //Since function type is first word, force the first char to upper case to match style of other error messages
+                    let properFunctionName = functionType.text[0].toUpperCase() + functionType.text.substring(1);
+                    //this error should not prevent parsing, but should still show up in the list of errors
+                    try {
+                        addError(new ParseError(name, `${properFunctionName} name cannot end with type designator '${lastChar}'`));
+                    } catch (e) { }
```

Comment:
Oh dang, I think this is the first place where we're emitting an error but still producing an AST node.  Looks really neat!  I'll try to think of a better way to handle this case than a `try/catch` around `addError`.  Or perhaps it just doesn't throw anymore?  We'll see!

Reply:
Ok, so this comment got me thinking about a way to remove the try catch. I just pushed some new code where I removed the throw inside of `addError`, and now, the code _calling_ `addError` is responsible for throwing. 

I also moved the creation of `ParserError` into of `addError`. I think this cuts down on the verbosity of throwing errors because you don't need to new up a new `ParseError` every time.

These changes help readability because an explicit `throw` shows that the code will short-circuit at the call site of `addError`. 

Let me know what you think. I can back it out if you don't like it.

Reply:
Thanks for taking a stab at that - I definitely like the explicit `throw` you've added!  At some level I liked `new`ing a `ParseError` for that same level of explicitness (explicitacity?), but since every one wound up being a `ParseError` it seems better this way.  Good call!